File: RiskAndroid/src/main/java/cc/game/android/risk/PlayerChooserDialog.java
package cc.game.android.risk;

import android.content.DialogInterface;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.ListView;
import android.widget.Toast;
import android.widget.ToggleButton;

import java.util.ArrayList;
import java.util.List;

import cc.lib.game.Utils;
import cc.lib.risk.Army;
import cc.lib.risk.RiskPlayer;

/**
 * Created by Chris Caron on 9/17/21.
 */
public class PlayerChooserDialog extends BaseAdapter implements DialogInterface.OnClickListener {

    final RiskActivity context;

    static class PL {
        boolean checked;
        boolean robot;
        Army army;

        PL(Army army) {
            this.army = army;
        }
    }

    List<PL> players = new ArrayList<>();

    PlayerChooserDialog(RiskActivity context) {
        this.context = context;
        for (Army army : Army.values()) {
            if (army != Army.NEUTRAL)
                players.add(new PL(army));
        }
        ListView lv = new ListView(context);
        lv.setAdapter(this);
        context.newDialogBuilder().setTitle("Choose Players")
                .setView(lv)
                .setNegativeButton(R.string.popup_button_cancel, null)
                .setPositiveButton(R.string.popup_button_ok, this).show();
    }

    @Override
    public int getCount() {
        return players.size();
    }

    @Override
    public Object getItem(int position) {
        return null;
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View view, ViewGroup parent) {
        if (view == null) {
            view = View.inflate(context, R.layout.player_list_item, null);
        }

        CheckBox cb = view.findViewById(R.id.check_box);
        ToggleButton tb = view.findViewById(R.id.toggle_button);

        cb.setOnCheckedChangeListener(null);
        tb.setOnCheckedChangeListener(null);

        PL pl = players.get(position);
        cb.setText(pl.army.name());
        cb.setChecked(pl.checked);
        tb.setChecked(pl.robot);

        cb.setOnCheckedChangeListener((buttonView, isChecked) -> pl.checked = !pl.checked);
        tb.setOnCheckedChangeListener((buttonView, isChecked) -> pl.robot = !pl.robot);

        return view;
    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        if (Utils.count(players, pl->pl.checked) < 2) {
            Toast.makeText(context, "Not enough players", Toast.LENGTH_LONG).show();
            return;
        }

        List<RiskPlayer> p = new ArrayList<>();
        for (PL pl : players) {
            if (pl.checked) {
                if (pl.robot) {
                    p.add(new RiskPlayer(pl.army));
                } else {
                    p.add(new UIRiskPlayer(pl.army));
                }
            }
        }

        context.startGame(p);
    }
}

File: RiskAndroid/src/main/java/cc/game/android/risk/PlayerChooserDialog.kt
package cc.game.android.risk

import android.content.DialogInterface
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import cc.game.android.risk.databinding.PlayerListItemBinding
import cc.lib.android.LinearRecyclerView
import cc.lib.risk.Army
import cc.lib.risk.RiskPlayer
import cc.lib.risk.UIRiskPlayer

class PL(var army: Army) {
	var checked = false
	var robot = false
}

class PlayerViewHolder(val binding: PlayerListItemBinding) : RecyclerView.ViewHolder(binding.root)

/**
 * Created by Chris Caron on 9/17/21.
 */
class PlayerChooserDialog internal constructor(val context: RiskActivity) : RecyclerView.Adapter<PlayerViewHolder>(), DialogInterface.OnClickListener {

	var players: MutableList<PL> = ArrayList()

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int):
		PlayerViewHolder = PlayerViewHolder(PlayerListItemBinding.inflate(LayoutInflater.from(parent.context)))

	override fun onBindViewHolder(holder: PlayerViewHolder, position: Int) {
		holder.binding.pl = players[position]
		holder.binding.lifecycleOwner = context
	}

	override fun getItemCount(): Int = players.size

	override fun onClick(dialog: DialogInterface, which: Int) {
		if (players.count { it.checked } < 2) {
			Toast.makeText(context, "Not enough players", Toast.LENGTH_LONG).show()
			return
		}

		players.filter { it.checked }.map {
			if (it.robot)
				RiskPlayer(it.army)
			else
				UIRiskPlayer(it.army)
		}.apply {
			context.game.startGame(this)
		}
	}

	init {
		Army.choices().forEach {
			players.add(PL(it))
		}
		context.newDialogBuilder().setTitle("Choose Players")
			.setView(LinearRecyclerView(context).also {
				it.adapter = this
			})
			.setNegativeButton(R.string.popup_button_cancel, null)
			.setPositiveButton(R.string.popup_button_ok, this).show()
	}
}